src/integrationtests/java/com/aws/greengrass/integrationtests/extensions/IntegrationTestExtension.java

```diff
@@ -0,0 +1,217 @@
+                invocation.skip();
+                return;
+            }
+            startMqtt5Broker();
+        } else if (extensionContext.getRequiredTestMethod().getAnnotation(TestWithAllBrokers.class) != null
+                || extensionContext.getRequiredTestMethod().getAnnotation(TestWithMqtt3Broker.class) != null
+                || extensionContext.getRequiredTestMethod().getAnnotation(TestWithMqtt5Broker.class) != null) {
+            fail("Please add an argument of type " + Broker.class.getSimpleName()
+                    + " to parameterized test " + extensionContext.getRequiredTestMethod().getName()
+                    + ". It's required by IntegrationTestExtension to determine which broker to spin up");
+        }
+
+        String configFile = getConfigFile(extensionContext);
+        if (configFile == null) {
+            invocation.proceed();
+            return;
+        }
+
+        // start kernel
+        startKernel(extensionContext, configFile);
+        // hack to set bridge to use mqtt5 broker after startup,
+        // since testcontainers picks a dynamic broker port
+        if (context.broker == Broker.MQTT5) {
+            pointBridgeToV5Broker();
+        }
+
+        invocation.proceed();
+    }
+
+    @Override
+    public void afterTestExecution(ExtensionContext extensionContext) {
+        if (kernel != null) {
+            kernel.shutdown();
+        }
+        if (v5Broker != null) {
+            v5Broker.stop();
+        }
+        if (v3Broker != null) {
+            v3Broker.stopServer();
+        }
+    }
+
+    private void initializeContext(ExtensionContext extensionContext) {
+        context = new IntegrationTestContext();
+        injectContext(extensionContext, context);
+
+        // relies on UniqueRootPathExtension being run before this extension via the IntegrationTest annotation
+        context.setRootDir(Paths.get(System.getProperty("root")));
+    }
+
+    private void startKernel(ExtensionContext extensionContext, String configFile) throws InterruptedException {
+        System.setProperty("aws.greengrass.scanSelfClasspath", "true");
+        kernel = new Kernel();
+        kernel.getContext().put(CertificateManager.class, mock(CertificateManager.class));
+        startKernelWithConfig(extensionContext, configFile);
+        context.setKernel(kernel);
+    }
+
+    private void startMqtt3Broker() throws IOException {
+        int brokerPort = 8883;
+        IConfig brokerConf = new MemoryConfig(new Properties());
+        brokerConf.setProperty(BrokerConstants.PORT_PROPERTY_NAME, String.valueOf(brokerPort));
+        v3Broker = new Server();
+        v3Broker.startServer(brokerConf);
+        context.setBrokerHost("localhost");
+        context.setBrokerPort(brokerPort);
+        context.setBroker(Broker.MQTT3);
+    }
+
+    private void startMqtt5Broker() {
+        v5Broker = new HiveMQContainer(
+                DockerImageName.parse("hivemq/hivemq-ce").withTag("2021.3"))
```

Comment:
use the latest? They should have something from 2023 or at least 2022

Reply:
thats what i get for pulling from docs w/o verifying 😅